File: ts/earthbucks-lib/src/ebx-error.ts
import { Result, Ok, Err } from "earthbucks-opt-res/src/lib.js";
import { Option, Some, None } from "earthbucks-opt-res/src/lib.js";

export abstract class EbxError extends Error {
  constructor() {
    super();
  }
}

export class GenericError extends EbxError {
  constructor(
    public source: Option<EbxError>,
    public message: string,
  ) {
    super();
  }

  toString(): string {
    return `ebx error: ${this.message}`;
  }
}

export class InvalidSizeError extends EbxError {
  constructor(public source: Option<EbxError>) {
    super();
  }

  toString(): string {
    return `invalid size`;
  }
}

export class NotEnoughDataError extends EbxError {
  constructor(public source: Option<EbxError>) {
    super();
  }

  toString(): string {
    return `not enough bytes in the buffer to read`;
  }
}

export class TooMuchDataError extends EbxError {
  constructor(public source: Option<EbxError>) {
    super();
  }

  toString(): string {
    return `too many bytes in the buffer to read`;
  }
}

export class NonMinimalEncodingError extends EbxError {
  constructor(public source: Option<EbxError>) {
    super();
  }

  toString(): string {
    return `non-minimal encoding`;
  }
}

export class InsufficientPrecisionError extends EbxError {
  constructor(public source: Option<EbxError>) {
    super();
  }

  toString(): string {
    return `number too large to retain precision`;
  }
}

export class InvalidOpcodeError extends EbxError {
  constructor(public source: Option<EbxError>) {
    super();
  }

  toString(): string {
    return `invalid opcode`;
  }
}

export class InvalidHexError extends EbxError {
  constructor(public source: Option<EbxError>) {
    super();
  }

  toString(): string {
    return `invalid hex`;
  }
}

export class InvalidEncodingError extends EbxError {
  constructor(public source: Option<EbxError>) {
    super();
  }

  toString(): string {
    return `invalid encoding`;
  }
}

export class InvalidKeyError extends EbxError {
  constructor(public source: Option<EbxError>) {
    super();
  }

  toString(): string {
    return `invalid key`;
  }
}

export class InvalidChecksumError extends EbxError {
  constructor(public source: Option<EbxError>) {
    super();
  }

  toString(): string {
    return `invalid checksum`;
  }
}

File: rs/earthbucks_lib/src/ebx_error.rs
use std::{error::Error, fmt};

#[derive(Debug)]
pub enum EbxError {
    TooLittleDataError { source: Option<Box<EbxError>> },
    TooMuchDataError { source: Option<Box<EbxError>> },
    NonMinimalEncodingError { source: Option<Box<EbxError>> },
    InsufficientPrecisionError { source: Option<Box<EbxError>> },
    InvalidOpcodeError { source: Option<Box<EbxError>> },
    InvalidHexError { source: Option<Box<EbxError>> },
    InvalidEncodingError { source: Option<Box<EbxError>> },
    InvalidKeyError { source: Option<Box<EbxError>> },
    InvalidChecksumError { source: Option<Box<EbxError>> },
}

impl Error for EbxError {
    fn source(&self) -> Option<&(dyn Error + 'static)> {
        match self {
            EbxError::TooLittleDataError { source } => {
                source.as_deref().map(|s| s as &(dyn Error + 'static))
            }
            EbxError::TooMuchDataError { source } => {
                source.as_deref().map(|s| s as &(dyn Error + 'static))
            }
            EbxError::NonMinimalEncodingError { source } => {
                source.as_deref().map(|s| s as &(dyn Error + 'static))
            }
            EbxError::InsufficientPrecisionError { source } => {
                source.as_deref().map(|s| s as &(dyn Error + 'static))
            }
            EbxError::InvalidOpcodeError { source } => {
                source.as_deref().map(|s| s as &(dyn Error + 'static))
            }
            EbxError::InvalidHexError { source } => {
                source.as_deref().map(|s| s as &(dyn Error + 'static))
            }
            EbxError::InvalidEncodingError { source } => {
                source.as_deref().map(|s| s as &(dyn Error + 'static))
            }
            EbxError::InvalidKeyError { source } => {
                source.as_deref().map(|s| s as &(dyn Error + 'static))
            }
            EbxError::InvalidChecksumError { source } => {
                source.as_deref().map(|s| s as &(dyn Error + 'static))
            }
        }
    }
}

impl fmt::Display for EbxError {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            EbxError::TooLittleDataError { .. } => {
                write!(f, "not enough bytes in the buffer to read")
            }
            EbxError::TooMuchDataError { .. } => {
                write!(f, "too many bytes in the buffer to read")
            }
            EbxError::NonMinimalEncodingError { .. } => {
                write!(f, "non-minimal encoding")
            }
            EbxError::InsufficientPrecisionError { .. } => {
                write!(f, "number too large to retain precision")
            }
            EbxError::InvalidOpcodeError { .. } => {
                write!(f, "invalid opcode")
            }
            EbxError::InvalidHexError { .. } => {
                write!(f, "invalid hex")
            }
            EbxError::InvalidEncodingError { .. } => {
                write!(f, "invalid encoding")
            }
            EbxError::InvalidKeyError { .. } => {
                write!(f, "invalid private key")
            }
            EbxError::InvalidChecksumError { .. } => {
                write!(f, "invalid checksum")
            }
        }
    }
}
